### ISSUE TITLE
Changes to HEAD not detected when running on a background thread

### ISSUE THREAD

**Issue by rytmis** [libgit2/libgit2sharp#1269]
Caveat: This may simply be a question of me not understanding how to use the library. In fact, that's the likely explanation. Anyway.

I'm trying to build a Git GUI client for Windows, with the principle that all expensive IO should be done on a background thread, to keep the UI responsive. The repository is initialized within a Task.Run block and everything works out nicely... mostly. Here's the weird part, though: I wanted to detect changes that happen outside my application, so I added a FileSystemWatcher on the repository directory. It has two things to do: first, check whether the current HEAD has changed, in order to trigger a recalculation of the commit graph. Second, if it hasn't, it checks the state of the working directory to display those changes only. 

The second path works exactly as it should, and my app detects simple file changes to the work dir just fine. However, if I use another Git client to switch branches, the app never notices that.

The interesting part of my code looks like this:

```csharp
if (uncommitted.ParentCommitId != Repository.Head.Tip.Id) {
  await LoadCommitsAsync();
  return;
}
```

Here, `uncommitted` is a view model that represents the working directory changes. This branch is _never_ executed -- unless I'm running in a debugger, and I _explicitly evaluate `Repository.Head.Tip.Id`_, in which case the execution flow _does_ go to that branch and the commit graph _is_ redrawn.

When evaluating `Repository.Head.Tip.Id`, I have to tell the debugger to let all threads run -- this, I assume, is a side effect of the properties being backed by `Lazy<T>` -- which sort of leads me to think there's some sort of thread affinity in play here. The repository is initialized on a background thread (Task.Run), and the example code runs in response to any one of the events on `FileSystemWatcher`.

Am I going crazy or am I just dumb? :)

**Comment by carlosmn** [comment]
Without seeing what code is running, we can't tell what's happening. As long as you're not using objects from multiple threads at the same time, it should work just fine.

**Comment by carlosmn** [comment]
Without more information or a reproducible case we can't do anything from this project, so keeping this open doesn't help. Feel free to reopen if you can come up with a reproducible case we can try out.

**Comment by rytmis** [comment]
Yeah, sorry about not responding faster. I've been too busy to cook up a simpler repro. I'll try to do that as soon as I can. Thanks for your patience.

**Comment by rytmis** [comment]
For the sake of completeness, I finally got around to looking at my pet project again. After adding a wee bit of debug logging, the problem was obvious and entirely my own. I simply jumped to the wrong conclusion because everything seemed to work when hitting a breakpoint but not otherwise. Long story short, I neglected to account for the fact that I'd detect and react to the _first_ external change, and checking to see whether the current HEAD had changed was something you would obviously want to do only after the external changes had actually completed. In the debugger, that would always be the case, but outside of it, practically never. :)
